keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,411 @@
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+  - name: worker
+    replicas: 2
+    template:
+      spec:
+        backoffLimit: 0 
+        completions: 2
+        parallelism: 2
+        template:
+          spec:
+            containers:
+            - name: worker
+              image: bash:latest
+              command:
+              - bash
+              - -xc
+              - |
+                sleep 10
+```
+
+### Risks and Mitigations
+
+To avoid the JobSet become a half-baked workflow manager, we will not allow for starting particular replicated jobs.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+### API Proposal
+  
+This is an annotation added to a JobSet when `InOrder` is used.
+It will track the next ReplicatedJob that is to be started/resumed.
+
+```golang
+const (
+ ReplicatedJobStart string = "jobset.sigs.k8s.io/startup-policy-next-to-start"
+)
+
+type StartupPolicyOrderOptions string
+
+const (
+ // This is the default setting
+ // AnyOrder means that we will start jobs without any specific order.
+ AnyOrder StartupPolicyOrderOptions = "AnyOrder"
+ // InOrder starts the jobs in order that they are listed.
+ InOrder StartupPolicyOrderOptions = "InOrder"
+)
+
+type StartupPolicy struct {
+ // StartupPolicyOrder is an enum of different options for StartupPolicy
+ // AnyOrder means to start replicated jobs in any order.
+ // InOrder means to start them as they are listed in the JobSet.
+ // +kubebuilder:validation:Enum=AnyOrder;InOrder
+ StartupPolicyOrder StartupPolicyOrderOptions `json:"startupPolicyOrder"`
+}
+```
+
+### Implementation
+
+We have three cases where startupPolicy applies:
+
+- when the jobset is first created
+- when the jobset is restarted after a failure (with restart or recreate all failure policy)
+- when the jobset is resumed after suspension
+
+In cases 1 and 2, the operator controls the order of creating the jobs, and so it creates the replicatedJob only when its turn comes in the startup sequence.
+In case 3, the operator controls the order of unsuspending the jobs, and so it unsuspends the replicatedJob only when its turn comes in the startup sequence.
+
+Since each start/resume of a replicated job will trigger a reconcile, we have to track which replicated job is to start/resume.
+The annotation `jobset.sigs.k8s.io/startup-policy-next-to-start` will be updated on each reconcile to reflect what is the next ReplicatedJob to resume.
+It is implied that the previous all of the jobs of the previous replicated job must be ready before this replicated job is started.
```

Comment:
not sure what you mean by implied, but this is something the controller need to explicitly check: start the next replicatedJob only when the previous one is ready.

Reply:
yes, the controller will check that. I said implied because we don't have any API or condition that says which job is starting (and we are implicitly waiting for all the jobs to be ready). So its implied in the sense that there is no explict API stating that we are waiting on this job.

The controller should look at the previous job and verify that it is all ready before it starts the next one.

Reply:
Why not just look at the statuses of all the Jobs in each reconcile? Caching the "next step" is probably going to lead to race conditions.

One corner case to consider is if a Job immediate terminates successfully. What would be the semantics in that case?

Reply:
In this case, the main problem is what replicated job do you want to start next? If replicated job a is all ready, then we should go to b. But looking at the status of all jobs means you may see that both b and c are not yet started but you aren't tracking which one is next to be started.

In the case of job terminate successfully wouldn't that be considered successful eventually in the status field of the jobset?

Reply:
> In this case, the main problem is what replicated job do you want to start next? If replicated job a is all ready, then we should go to b. But looking at the status of all jobs means you may see that both b and c are not yet started but you aren't tracking which one is next to be started.

Shouldn't the order determined by the order of the replicatedJobs in the spec?

Reply:
Yes, maybe we don't need the annotation after all. I think we have ready, suspended, failed and succeeded as status for ReplicatedJobs.

I think the main issue is on creation or restart, I am thinking you want to track which jobs were created/recreated in case that they are still waiting to be ready.

What I have to do is on `createJobs` I want to only create the jobs in order specified in the startup policy. But it make take a few reconciles for those jobs to start but I think we could look at the status to determine that.

@ahg-g wdyt?

Reply:
why is restart different? You should just iterate through all the jobs and see if they are running or not.

> In the case of job terminate successfully wouldn't that be considered successful eventually in the status field of the jobset?

That's a question for your design that should ultimately be reflected in the documentation. It's not necessarily obvious what the semantics are.

Reply:
Also what if the job has a TTLAfterFinished?

Reply:
Yes, we don't need to rely on an explicit state to track what to start next. But I think it is good UX to have a condition indicating that the JobSet is starting up `InOrder` and indicate the progress (as you mentioned, the reason perhaps indicates which replicatedJob we started last or will start next).

I am not sure I follow the question on job termination, if the job terminates, then we are done, why is this relevant to startup sequence? In the same vein, why is TTLAfterFinished relevant as well?

Reply:
IMO its not related to startup policy but its a fair point if we use `replicatedJobStatus` to track state. Since that relies on the Jobs being present in the apiserver, what happens if those jobs are removed from etcd?

If `ttlAfterFinished` is set to a low value (or a job finishes and is deleted quickly), then replicatedJobStatus will no longer detect the status of those replicated jobs. I could see a similar issue happening if `successPolicy` or `failurePolicy` is used. 

Now this seems to be an edge case because the answer I think of is don't set your `ttlAfterFinished` to a time that is less than a reconcile period..

Reply:
We need to have finalizers on the child jobs to make sure that their status is synced on the replicatedJob status before they get deleted. 

So I agree that we need to take a dependency on the replicatedJob status when determining if the jobs under a replicated job are ready or not. We can address the correctness of status in the face of deleted job objects using finalizers as a followup.

Reply:
I think this  belongs outside of StartupPolicy so I created https://github.com/kubernetes-sigs/jobset/issues/360 to document this.

Reply:
Yes, but there is still <<Unresolved>> tag in the text.

There are a bunch of suggestions as well that were not addressed yet.

Reply:
I address those and updated this.